### PR TITLE
Perfect square using binary search

### PR DESCRIPTION
**Pull request by IhebHaboubi**
### **Describe your change:**

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Review by rtvk007 — APPROVED** [review]

File: maths/perfect_square.py [hunk 1]
```diff
@@ -21,6 +21,36 @@ def perfect_square(num: int) -> bool:
...
+    >>> perfect_square_binary_search(9)
+    True
+    >>> perfect_square_binary_search(16)
+    True
+    >>> perfect_square_binary_search(1)
+    True
+    >>> perfect_square_binary_search(0)
+    True
+    >>> perfect_square_binary_search(10)
+    False
```
Comment (cclauss) [inline maths/perfect_square.py:39]: tests for -1, 1.1, "a", None, []?

**Review by swapnanildutta — APPROVED** [review]

File: maths/perfect_square.py (hunk 1 above)
Comment (IhebHaboubi) [reply maths/perfect_square.py:39]: should I first add input validation inside the function, or should I just include those tests ?

Comment (cclauss) [reply maths/perfect_square.py:39]: Write the test first and see how your function deals with unexpected input.  The trick is to not write validation code if you don't have to. If your function is already going to raise an IndexError, TypeError, ValueError, etc. then you have no work to do. Just look for cases where your code would silently fail or would deliver nonsense results.

Comment (IhebHaboubi) [reply maths/perfect_square.py:39]: Got it.Thank you so much ! 
Things should be good now.

**Review by cclauss — APPROVED** [review]

**Pull request merged**